### PR TITLE
Add scaffold for new BASP implementation

### PR DESCRIPTION
**Pull request by Neverlord**
Note: BASP is *not* feature complete yet. However, the diff has already grown quite large and also includes refactoring of transport, etc.

I think it's best to merge this first chunk of BASP and then add the missing pieces in smaller PRs later. @jakobod, @josephnoir I've added you both for review, because I think it's important that we're on the same page at least for the changes to non-BASP classes. Since @josephnoir is familiar with the previous BASP implementation, I think it's ok if @jakobod only skims the implementation details of the `basp::application`. 🙂

### CODE REVIEW

File: libcaf_net/src/ec.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+const char* ec_names[] = {
+  "none",
+  "invalid_magic_number",
+  "unexpected_number_of_bytes",
+  "unexpected_payload",
+  "missing_payload",
+  "illegal_state",
+  "invalid_handshake",
+  "missing_handshake",
+  "version_mismatch",
```
Comment (josephnoir) [inline libcaf_net/src/ec.cpp]: I think `"unexpected handshake"` is missing in the list.

File: libcaf_net/src/message_type.cpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+#include "caf/string_view.hpp"
+
+namespace caf {
+namespace net {
+namespace basp {
+
+namespace {
+
+string_view message_type_names[] = {
```
Comment (josephnoir) [inline libcaf_net/src/message_type.cpp]: Similar enums above (`ec`, `connection_state`) use a `const char*` for this list instead of a `string_view`.

File: libcaf_net/caf/net/basp/header.hpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+  header& operator=(const header&) noexcept = default;
+
+  int compare(header other) const noexcept;
+
+  /// @pre `bytes.size() == header_size`
+  static header from_bytes(span<const byte> bytes);
+};
+
+/// Size of a BASP header in serialized form
```
Comment (josephnoir) [inline libcaf_net/caf/net/basp/header.hpp]:
```suggestion
/// Size of a BASP header in serialized form.
```

File: libcaf_net/src/application.cpp [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+  serializer_impl<buffer_type> sink{system(), buf_};
+  if (auto err = sink(path)) {
+    CAF_LOG_ERROR("unable to serialize path");
+    return;
+  }
+  auto req_id = next_request_id_++;
+  auto hdr = to_bytes(header{message_type::resolve_request,
+                             static_cast<uint32_t>(buf_.size()), req_id});
+  if (auto err = write_packet(hdr, buf_)) {
+    CAF_LOG_ERROR("unable to serialize path");
```
Comment (josephnoir) [inline libcaf_net/src/application.cpp]:
```suggestion
    CAF_LOG_ERROR("unable to write resolve request header");
```

File: libcaf_net/src/application.cpp [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+    return err;
+  if (aid == 0) {
+    i->second.deliver(strong_actor_ptr{nullptr}, std::move(ifs));
+    return none;
+  }
+  i->second.deliver(proxies_->get_or_put(peer_id_, aid), std::move(ifs));
+  return none;
+}
+
+error application::generate_handshake() {
```
Comment (josephnoir) [inline libcaf_net/src/application.cpp]: Not sure if we ever want to retransmit handshakes. If so, it might be a good idea to clear `buf_` before filling it.

File: libcaf_net/caf/net/basp/message_type.hpp [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+/// @addtogroup BASP
+
+/// Describes the first header field of a BASP message and determines the
+/// interpretation of the other header fields.
+enum class message_type : uint8_t {
+  /// Sends supported BASP version and node information to the server.
+  ///
+  /// ![](client_handshake.png)
+  handshake = 0x00,
```
Comment (jakobod) [inline libcaf_net/caf/net/basp/message_type.hpp]: Is there a specific reason you are using hexadecimal here instead of decimal like you do in the other enum definitions?

File: libcaf_net/caf/net/stream_transport.hpp [hunk 7]
```diff
@@ -61,18 +66,35 @@ class stream_transport {
...
+    return *this;
+  }
+
+  endpoint_manager& manager() {
+    return *manager_;
+  }
+
   // -- member functions -------------------------------------------------------
 
   template <class Parent>
```
Comment (jakobod) [inline libcaf_net/caf/net/stream_transport.hpp]: Is the `Parent` template and parameter still needed even though you the getters for `endpoint_managers` and so on above?

Comment (jakobod) [reply libcaf_net/caf/net/stream_transport.hpp]: Especially if you are using the specific type `endpoint_manager*` on line 237. This template seems especially odd..

File: libcaf_net/caf/net/endpoint_manager_impl.hpp [hunk 8]
```diff
@@ -48,10 +52,18 @@ class endpoint_manager_impl : public endpoint_manager {
...
 
+  application_type& application() {
+    return transport_.application();
+  }
+
   transport_type& transport() {
     return transport_;
   }
 
+  endpoint_manager_impl& manager() {
```
Comment (jakobod) [inline libcaf_net/caf/net/endpoint_manager_impl.hpp]: What is this getter for?

File: libcaf_net/caf/net/transport_worker.hpp [hunk 9]
```diff
@@ -65,20 +81,16 @@ class transport_worker {
     application_.write_message(decorator, std::move(msg));
   }
 
-  template <class Parent>
-  void write_packet(Parent& parent, span<const byte> header,
```
Comment (jakobod) [inline libcaf_net/caf/net/transport_worker.hpp]: Why did you delete this function? Does the application need to get the `id_type` and pass it on to the transport itself?

File: libcaf_net/caf/net/basp/application.hpp [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+  template <class Parent>
+  error write_message(Parent& parent,
+                      std::unique_ptr<endpoint_manager::message> ptr) {
+    // TODO: avoid extra copy of the payload
+    buf_.clear();
+    serializer_impl<buffer_type> sink{system(), buf_};
+    const auto& src = ptr->msg->sender;
+    const auto& dst = ptr->receiver;
+    if (dst == nullptr) {
+      // TODO: valid?
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_net/caf/net/basp/application.hpp]: If we have a need for system messages then this could be one way of sending them.

File: libcaf_net/caf/net/basp/connection_state.hpp [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+namespace net {
+namespace basp {
+
+/// @addtogroup BASP
+
+/// Stores the state of a connection in a `basp::application`.
+enum class connection_state {
+  /// Indicates that we successfully checked the magic number and now wait for
+  /// the handshake header.
+  await_handshake_header,
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_net/caf/net/basp/connection_state.hpp]: Sorry for the confusion. There is no magic number. I've started implementing BASP with a magic number but it only makes things complicated.

File: libcaf_net/src/application.cpp [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+    case message_type::resolve_response:
+      return handle_resolve_response(write_packet, hdr, payload);
+    case message_type::heartbeat:
+      return none;
+    default:
+      return ec::unimplemented;
+  }
+}
+
+error application::handle_handshake(write_packet_callback&, header hdr,
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_net/src/application.cpp]: Good catch!

File: libcaf_net/src/message_type.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_net/src/message_type.cpp]: I think we should actually get rid of all the `const char*` stuff and make consistent use of `string_view` instead.

File: libcaf_net/caf/net/stream_transport.hpp [hunk 13]
```diff
@@ -82,8 +104,7 @@ class stream_transport {
       CAF_LOG_DEBUG(CAF_ARG(len) << CAF_ARG(handle_.id) << CAF_ARG(*num_bytes));
       collected_ += *num_bytes;
       if (collected_ >= read_threshold_) {
-        auto decorator = make_write_packet_decorator(*this, parent);
-        worker_.handle_data(decorator, read_buf_);
+        worker_.handle_data(*this, read_buf_);
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_net/caf/net/stream_transport.hpp]: The decorator always calls `set_timeout` directly on the parent. Do you have a use case where we need to delegate `set_timeout` via decorator?

File: libcaf_net/src/application.cpp [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+                                    byte_span payload) {
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_net/src/application.cpp]: > I don't think this is used.

Is is now. 🙂

File: libcaf_net/caf/net/basp/message_type.hpp (hunk 6 above)
Comment (Neverlord) [reply libcaf_net/caf/net/basp/message_type.hpp]: Not really. Switched to decimal.

File: libcaf_net/caf/net/stream_transport.hpp (hunk 7 above)
Comment (Neverlord) [reply libcaf_net/caf/net/stream_transport.hpp]: Yes, because `Parent` is the actual type, i.e., `endpoint_manager_impl<...>` and we might want to wrap the transport one day.

File: libcaf_net/caf/net/endpoint_manager_impl.hpp (hunk 8 above)
Comment (Neverlord) [reply libcaf_net/caf/net/endpoint_manager_impl.hpp]: We need to have a reference/pointer to the manager for creating proxies.

File: libcaf_net/caf/net/transport_worker.hpp (hunk 9 above)
Comment (Neverlord) [reply libcaf_net/caf/net/transport_worker.hpp]: The only job of the decorator is to add this ID. That's why I changed `write_packet` in the decorator to avoid having all worker classes implement virtually identical `write_packet` functions. Here's the new code in the decorator:

```c++
  template <class... Ts>
  void write_packet(span<const byte> header, span<const byte> payload,
                    Ts&&... xs) {
    parent_.write_packet(header, payload, std::forward<Ts>(xs)...,
                         object_.id());
  }
```

File: libcaf_net/caf/net/stream_transport.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (josephnoir) [reply libcaf_net/caf/net/stream_transport.hpp]: I'm good with that. Just had it in the back of my head as a reason why we originally did this ...  that might have been a few iterations ago. ^^

File: libcaf_net/src/message_type.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (josephnoir) [reply libcaf_net/src/message_type.cpp]: Let's do that.

File: libcaf_net/src/ec.cpp [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+  "invalid_magic_number",
+  "unexpected_number_of_bytes",
+  "unexpected_payload",
+  "missing_payload",
+  "illegal_state",
+  "invalid_handshake",
+  "missing_handshake",
+  "unexpected_handshake",
+  "version_mismatch",
+  "unimplemented = 10",
```
Comment (josephnoir) [inline libcaf_net/src/ec.cpp]:
```suggestion
  "unimplemented",
```

Comment (Neverlord) [reply libcaf_net/src/ec.cpp]: I sneaked that in there to test your vigilance. 😏

File: libcaf_net/caf/net/stream_transport.hpp (hunk 7 above)
Comment (jakobod) [reply libcaf_net/caf/net/stream_transport.hpp]: Oh yes, I see that! 👍

File: libcaf_net/caf/net/transport_worker.hpp (hunk 9 above)
Comment (jakobod) [reply libcaf_net/caf/net/transport_worker.hpp]: This does indeed seem cleaner. 👍

Comment (josephnoir) [reply libcaf_net/caf/net/transport_worker.hpp]: Constant vigilance! 🧙‍♂️

**Comment by josephnoir** [comment]
Ok, one last thing. How does the separate case or connection state `await_handshake_payload` help us? From the code it seems that we could do without it and move to `await_payload` directly. `handle` (in line 145) could then dispatch to `handle_handshake`. The difference might be that `connection_state::await_handshake_payload` does not check the payload size of the header.

**Comment by Neverlord** [comment]
> Ok, one last thing. How does the separate case or connection state await_handshake_payload help us?

This state allows us to dispatch to `application::handle_handshake` only *once* and raise an error for repeated handshakes otherwise. Without that state, we can only dispatch on `hdr.type` and maybe do some guess work (check whether `peer_id_` is default-constructed or something along those lines). However, having this additional state gives us a very clean FSM that runs `handle_handshake` exactly once at the beginning and prohibits handshakes after the connection is established.

**Comment by josephnoir** [comment]
Ah, because the case for `connection_state::await_header` simply processes the header of a repeated handshake, but handling the payload would then fail.
